src/jetstream/repository/apikeys/psql_apikeys.go

```diff
@@ -0,0 +1,184 @@
+	// Iterating over the struct to ensure that all of the queries are updated
+	v := reflect.ValueOf(sqlQueries)
+	for i := 0; i < v.NumField(); i++ {
+		q := v.Field(i).Interface().(string)
+
+		reflect.
+			ValueOf(&sqlQueries).
+			Elem().
+			FieldByIndex([]int{i}).
+			SetString(
+				datastore.ModifySQLStatement(q, databaseProvider),
+			)
+	}
+}
+
+// AddAPIKey - Add a new API key to the datastore.
+func (p *PgsqlAPIKeysRepository) AddAPIKey(userID string, comment string) (*interfaces.APIKey, error) {
+	log.Debug("AddAPIKey")
+
+	var err error
+
+	// Validate args
+	if len(comment) > 255 {
+		msg := "comment maximum length is 255 characters"
+		log.Debug(msg)
+		err = errors.New(msg)
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	randomBytes, err := crypto.GenerateRandomBytes(48)
+	if err != nil {
+		return nil, err
+	}
+
+	keyGUID := uuid.NewV4().String()
+	keySecret := base64.URLEncoding.EncodeToString(randomBytes)
+
+	err = execQuery(p, sqlQueries.InsertAPIKey, keyGUID, keySecret, userID, comment)
+	if err != nil {
+		return nil, fmt.Errorf("AddAPIKey: %v", err)
+	}
+
+	apiKey := &interfaces.APIKey{
+		GUID:     keyGUID,
+		Secret:   keySecret,
+		UserGUID: userID,
+		Comment:  comment,
+	}
+
+	return apiKey, err
+}
+
+// GetAPIKeyBySecret - gets user ID for an API key
+func (p *PgsqlAPIKeysRepository) GetAPIKeyBySecret(keySecret string) (*interfaces.APIKey, error) {
+	log.Debug("GetAPIKeyBySecret")
+
+	var apiKey interfaces.APIKey
+
+	err := p.db.QueryRow(sqlQueries.GetAPIKeyBySecret, keySecret).Scan(
+		&apiKey.GUID,
+		&apiKey.UserGUID,
+		&apiKey.Comment,
+		&apiKey.LastUsed,
+	)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return &apiKey, nil
+}
+
+// ListAPIKeys - list API keys for a given user GUID
+func (p *PgsqlAPIKeysRepository) ListAPIKeys(userID string) ([]interfaces.APIKey, error) {
+	log.Debug("ListAPIKeys")
+
+	rows, err := p.db.Query(sqlQueries.ListAPIKeys, userID)
+	if err != nil {
+		log.Errorf("unable to list API keys: %v", err)
+		return nil, err
+	}
+
+	result := []interfaces.APIKey{}
+	for rows.Next() {
+		var apiKey interfaces.APIKey
+		err = rows.Scan(&apiKey.GUID, &apiKey.UserGUID, &apiKey.Comment, &apiKey.LastUsed)
+		if err != nil {
+			log.Errorf("Scan: %v", err)
+			return nil, err
+		}
+		result = append(result, apiKey)
+	}
+
+	return result, nil
+}
+
+// DeleteAPIKey - delete an API key identified by its GUID
+func (p *PgsqlAPIKeysRepository) DeleteAPIKey(userGUID string, keyGUID string) error {
+	log.Debug("DeleteAPIKey")
+
+	err := execQuery(p, sqlQueries.DeleteAPIKey, userGUID, keyGUID)
+	if err != nil {
+		return fmt.Errorf("DeleteAPIKey: %v", err)
+	}
+
+	return nil
+}
+
+// UpdateAPIKeyLastUsed - sets API key last_used field to current time
+func (p *PgsqlAPIKeysRepository) UpdateAPIKeyLastUsed(keyGUID string) error {
+	log.Debug("UpdateAPIKeyLastUsed")
+
+	err := execQuery(p, sqlQueries.UpdateAPIKeyLastUsed, time.Now(), keyGUID)
```

Comment:
It feels like this should be `time.Now().Unix()`. This would remove any localised issues when showing the date in the frontend. I've searched through the code for usages of `time.Now()` and this could be a wider problem, specifically when we check token expiration (CF in one timezone, Stratos in another).

Reply:
Ignore this! `time.Now()` does give the offset, this should be correctly interpreted by moment in the frontend